### PR TITLE
NoImprovement stopping criterion

### PR DESCRIPTION
**Pull request by leonlan**
# Changes
- `NoImprovement` stopping criterion. This criterion allows ALNS to iterate until the best solution has not improved for a user-specified number of iterations `n_iterations`.

### CODE REVIEW

File: alns/stop/NoImprovement.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        return self._n_iterations
+
+    def __call__(self, rnd: RandomState, best: State, current: State) -> bool:
+        if self._target is None or best.objective() < self._target:
+            self._target = best.objective()
+            self._counter = 0
+            return False
+
+        self._counter += 1
+        return self._counter > self.n_iterations
```
Comment (leonlan) [inline alns/stop/NoImprovement.py:40]: I'm slightly confused whether this should be `>` or `>=`. Maybe a fresh pair of eyes can help out..

Comment (N-Wouda) [reply alns/stop/NoImprovement.py:40]: After n iterations without improvement, the counter is n. If we want to stop at that point, we should check if the counter is `>=` to n.

This should be a test case, I think.

Comment (N-Wouda) [inline alns/stop/NoImprovement.py:40]: Question: does this work if the maximum number of iterations is zero? I fear not, but should we allow zero?

**Review by N-Wouda — CHANGES_REQUESTED** [review]

File: alns/stop/tests/test_no_improvement.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    """
+    NoImprovement(n_iterations)
+
+
+def test_n_iterations():
+    """
+    Test if the n_iterations parameter is correctly set.
+    """
+    stop = NoImprovement(3)
+    assert stop.n_iterations == 3
```
[the remark this answers is not in the document]
Comment (leonlan) [reply alns/stop/tests/test_no_improvement.py]: Writing `assert_` is not a habit yet, apologies!

File: alns/stop/NoImprovement.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+from alns.stop.StoppingCriterion import StoppingCriterion
+
+
+class NoImprovement(StoppingCriterion):
+    """
+    Criterion that stops if the best solution has not been improved
+    after a number of iterations.
+
+    The first iteration is used to initialize the criterion. Hence,
+    the algorithm will iterate at least n_iterations + 1 times.
```
[the remark this answers is not in the document]
Comment (leonlan) [reply alns/stop/NoImprovement.py]: This hack arose from the confusion related to choosing `>` and `>=` and covering the `max_iterations=0` case. I think the latest version addresses this now.

File: alns/stop/tests/test_no_improvement.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    for _ in range(100):
+        assert_(not stop(rnd, Zero(), Zero()))
+
+
+def test_after_n_iterations():
+    stop = NoImprovement(100)
+    rnd = RandomState()
+
+    for _ in range(100 + 1):
+        stop(rnd, Zero(), Zero())
```
[the remark this answers is not in the document]
Comment (leonlan) [reply alns/stop/tests/test_no_improvement.py]: Answer: The presented test only tests after the maximum number of iterations; another test called`test_before_n_iterations` covered the requested test case.

See newest version, which combines the two.

File: alns/stop/NoImprovement.py (hunk 1 above)
Comment (leonlan) [reply alns/stop/NoImprovement.py]: It did cover the zero case, albeit creating confusing variable names as you might have noticed.

The newest version should have fixed this.

File: alns/stop/tests/test_no_improvement.py [hunk 5]
```diff
@@ -39,40 +52,42 @@ def test_first_iteration():
...
 
-    for _ in range(n_iterations):
-        assert_(not stop(rnd, Two(), Zero()))
+    for _ in range(5):
+        assert_(not stop(rnd, One(), One()))
 
-    assert not stop(rnd, One(), Zero())
-    assert stop._counter == 0
+    assert_(not stop(rnd, Zero(), One()))
+    assert_equal(stop._counter, 0)
```
Comment (N-Wouda) [inline alns/stop/tests/test_no_improvement.py:93]: Refactor: we should not care if the `NoImprovement` class implements its behaviour using a counter or not: we only care that after a given number of iterations without improvement, it provides a stop signal (and not in all other occasions).

I suggest we remove this final test.

File: alns/stop/tests/test_no_improvement.py [hunk 6]
```diff
@@ -7,29 +7,42 @@
...
+    """
+    Test if setting max_iterations to zero stops when a non-improving
+    best solution has been found.
+    """
+    stop = NoImprovement(0)
+    rnd = RandomState()
+
+    assert_(not stop(rnd, One(), Zero()))
+    assert_(not stop(rnd, Zero(), Zero()))
+    assert_(stop(rnd, Zero(), Zero()))
```
Comment (N-Wouda) [inline alns/stop/tests/test_no_improvement.py:45]: Issue: this feels wrong, in a sense. No improving solution can be found in no iterations. So for zero iterations we should always stop.

I suggest we fix this by implementing `__call__` slightly differently.

File: alns/stop/NoImprovement.py [hunk 7]
```diff
@@ -10,25 +10,22 @@ class NoImprovement(StoppingCriterion):
...
+        self._max_iterations = max_iterations
         self._target: Optional[float] = None
 
     @property
-    def n_iterations(self) -> int:
-        return self._n_iterations
+    def max_iterations(self) -> int:
+        return self._max_iterations
 
     def __call__(self, rnd: RandomState, best: State, current: State) -> bool:
```
Comment (N-Wouda) [inline alns/stop/NoImprovement.py:30]: Looking at this method, I propose the following slight change:
```python
def __call__(...):
  if self._target is None or best.objective() < self._target:
    self._target = best.objective()
    self._counter = 0
  else:
    self._counter += 1

  return self._counter >= self.max_iterations
```
This ensures we always perform the test against max iterations.

File: alns/stop/NoImprovement.py [hunk 8]
```diff
@@ -10,25 +10,22 @@ class NoImprovement(StoppingCriterion):
...
-    def __init__(self, n_iterations: int):
-        if n_iterations < 0:
-            raise ValueError("n_iterations < 0 not understood.")
+    def __init__(self, max_iterations: int):
+        if max_iterations < 0:
+            raise ValueError("max_iterations < 0 not understood.")
 
-        self._n_iterations = n_iterations
+        self._max_iterations = max_iterations
         self._target: Optional[float] = None
```
Comment (N-Wouda) [inline alns/stop/NoImprovement.py:24]: Issue: AFAICT `self._counter` is never initialised before `__call__`.

**Review by N-Wouda — CHANGES_REQUESTED** [review]

File: alns/stop/NoImprovement.py (hunk 8 above)
Comment (N-Wouda) [reply alns/stop/NoImprovement.py:24]: [I had expected mypy to flag this (since we're adding an impromptu field not declared in `__init__`, but perhaps I have missed something]

File: alns/stop/tests/test_no_improvement.py (hunk 5 above)
Comment (leonlan) [reply alns/stop/tests/test_no_improvement.py:93]: Following the same principle, I also removed `test_counter_value_at_stop`.

File: alns/stop/NoImprovement.py (hunk 8 above)
Comment (leonlan) [reply alns/stop/NoImprovement.py:24]: > [I had expected mypy to flag this (since we're adding an impromptu field not declared in `__init__`, but perhaps I have missed something]

I expected that as well, but running mypy didn't flag anything.

File: alns/stop/tests/test_no_improvement.py [hunk 9]
```diff
@@ -66,28 +72,21 @@ def test_stop_after_max_iterations():
...
+    for _ in range(2):
+        assert_(not stop(rnd, One(), Zero()))
 
-    assert_(not stop(rnd, Zero(), One()))
-    assert_equal(stop._counter, 0)
+    for _ in range(100):
+        assert_(not stop(rnd, Zero(), Zero()))
+
+    for _ in range(100):
+        assert_(stop(rnd, Zero(), Zero()))
```
Comment (N-Wouda) [inline alns/stop/tests/test_no_improvement.py:92]: Comment: this is really minor, but since the function mentions `n` as an arbitrary parameter, maybe set this up using pytest's parametrize?

Comment (N-Wouda) [reply alns/stop/tests/test_no_improvement.py:92]: And similarly for the other test function with `n` in its name (just above this one).

File: alns/stop/tests/test_no_improvement.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    """
+    NoImprovement(max_iterations)
+
+
+def test_max_iterations():
+    """
+    Test if the max_iterations parameter is correctly set.
+    """
+    stop = NoImprovement(3)
+    assert_equal(stop.max_iterations, 3)
```
Comment (N-Wouda) [inline alns/stop/tests/test_no_improvement.py:32]: Idea: parametrize this as well.

Since this is functionally the same as the `does_not_raise` test above, maybe remove that test and use this instead?

File: alns/stop/tests/test_no_improvement.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    rnd = RandomState()
+
+    for _ in range(2):
+        assert_(not stop(rnd, One(), Zero()))
+
+    for _ in range(100):
+        assert_(not stop(rnd, Zero(), Zero()))
+
+    for _ in range(100):
+        assert_(stop(rnd, Zero(), Zero()))
```
Comment (N-Wouda) [inline alns/stop/tests/test_no_improvement.py:92]: Idea: parametrize both these test methods.

File: alns/stop/tests/test_no_improvement.py (hunk 10 above)
Comment (leonlan) [reply alns/stop/tests/test_no_improvement.py:32]: Question: when does it help to parametrize tests? I initially had them, but then removed them again. I thought that it was not necessary as long as the chosen parameters don't touch the edge cases. The edge cases are tested separately.

Comment (N-Wouda) [reply alns/stop/tests/test_no_improvement.py]: In this particular instance I just wanted to test that we get the initial input back from max_iterations, for some different values.

Normally I parametrise tests with tuples of `(*inputs, output)`. That's useful to cover all sorts of edge cases with a single test!

**Review by N-Wouda — APPROVED** [review]

**Comment by N-Wouda** [comment]
@leonlan looks good to me, thanks!
